### PR TITLE
[NAE-1599] Task-reffed fields are not placed in correct position

### PR DESCRIPTION
**Pull request by renczesstefan**
# Description

Order of data groups created from task reference fields was not correct on grid layout. It was given from the order that is defined in Petri net, but if there is a grid layout, task ref field can have specified Y positions.

Fixes [NAE-1599]

## Dependencies

No new dependencies were introduced.

### Third party dependencies

No new dependencies were introduced.

### Blocking Pull requests

There are no dependencies on other PR.

## How Has Been This Tested?

This was tested manually importing agreement.xml petri net and there is a unit test implemented for it.

- testTaskRefOrderOnGridLayout() [DataServiceTest.java](https://github.com/netgrif/application-engine/blob/release/6.0.4/src/test/groovy/com/netgrif/application/engine/workflow/DataServiceTest.groovy)

### Test Configuration

Netgrif Backend PR Config
| Name                | Tested on |
|---------------------| --------- |
| OS                  |   macOS Monterey 12.2.1        |
| Runtime             |  Java 11        |
| Dependency Manager  |  Maven 3.8.4         |
| Framework version   |  Spring Boot 2.6.2        |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes have been checked, personally or remotely, with @petrovicLubos  @minop 
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have resolved all conflicts with the target branch of the PR
- [x] I have updated and synced my code with the target branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes:
    - [x] Lint test
    - [x] Unit tests
    - [x] Integration tests
- [x] I have checked my contribution with code analysis tools:
    - [x] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [x] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [x] I have made corresponding changes to the documentation:
    - [x] Developer documentation
    - [x] User Guides
    - [x] Migration Guides

[NAE-1599]: https://netgrif.atlassian.net/browse/NAE-1599?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/netgrif/application/engine/workflow/service/DataService.java [hunk 1]
```diff
@@ -308,6 +308,13 @@ private GetDataGroupsEventOutcome getDataGroups(String taskId, Locale locale, Se
         return groups;
     }
 
+    private void resolveTaskRefOrderOnGrid(DataGroup dataGroup, Map<String, Field> dataFieldMap) {
+        if (dataGroup.getLayout() != null && dataGroup.getLayout().getType() == "grid") {
+            List<Field> dataGroupFields = dataGroup.getData().stream().map(dataFieldMap::get).sorted(Comparator.comparingInt(a -> a.getLayout().getY())).collect(Collectors.toList());
+            dataGroup.setData(new LinkedHashSet<>(dataGroupFields.stream().map(Field::getStringId).collect(Collectors.toList())));
```
Comment (minop) [inline src/main/java/com/netgrif/application/engine/workflow/service/DataService.java:314]: why iterate over the stream twice? The `map` operator can be chained after the sorting and the desired set of IDs can then be collected with `Collectors.toCollection( LinkedHashSet::new )`

File: src/test/resources/agreement.xml [hunk 2]
```diff
@@ -0,0 +1,657 @@
+<document xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://netgrif.github.io/petriflow/petriflow.schema.xsd">
```
Comment (minop) [inline src/test/resources/agreement.xml:1]: A smallest reproducible example should be used for the unit test. Also the unit test nets should be generic and should not contain any business logic.

**Review by minop — CHANGES_REQUESTED** [review]
Has it been agreed that the fix will be applied to backend rather than to frontend? The number of elements is usually small, but the function calls still add extra overhead to each backend call.

**Comment by minop** [comment]
> Has it been agreed that the fix will be applied to backend rather than to frontend? The number of elements is usually small, but the function calls still add extra overhead to each backend call.

This has been answered in a personal call. Frontend would need to reconstruct the entire nesting tree of the data groups, which is much easier to do on backend, where it can be created in the correct order with little effort

**Review by minop — APPROVED** [review]

**Review by machacjozef — APPROVED** [review]

**Pull request merged**